url/tests/unit.rs

```diff
@@ -1379,3 +1379,46 @@ fn serde_error_message() {
         r#"relative URL without a base: "§invalid#+#*Ä" at line 1 column 25"#
     );
 }
+
+#[test]
+fn test_can_be_a_base_with_set_path() {
+    use url::quirks;
+    let mut url = Url::parse("web+demo:/").unwrap();
+    assert!(!url.cannot_be_a_base());
+
+    url.set_path("//not-a-host");
+    assert_eq!(url.path(), "//not-a-host");
+
+    let segments: Vec<_> = url
+        .path_segments()
+        .expect("should have path segments")
+        .collect();
+
+    assert_eq!(segments, vec!["", "not-a-host"]);
+
+    assert_eq!(url.as_str(), "web+demo:/.//not-a-host");
+    quirks::set_hostname(&mut url, "test").unwrap();
+    assert_eq!(url.as_str(), "web+demo://test//not-a-host");
+    quirks::set_hostname(&mut url, "").unwrap();
+    assert_eq!(url.as_str(), "web+demo:/.//not-a-host");
```

Comment:
Actually, I was wrong about this. We are setting the host to "", not null (in fact you can't set it to null again).
So the result should be `web+demo:////not-a-host`

I've checked this with the reference URL parser
```
// Go to https://jsdom.github.io/whatwg-url/
// Open the devtools console and type in all the » lines
» import('./whatwg-url.mjs').then(m => whatwg = m.default)
Promise { <state>: "fulfilled", <value>: {…} }

» u = new whatwg.URL("web+demo:/")
Object { … }

» u.href
"web+demo:/"
» u.pathname = "//not-a-host"
"//not-a-host"
» u.href
"web+demo:/.//not-a-host"
» u.host = "host"
"host"
» u.href
"web+demo://host//not-a-host"
» u.host = ""
""
» u.href
"web+demo:////not-a-host" 
```